### PR TITLE
BUG Add wait between status checks AWS worker

### PR DESCRIPTION
**Pull request by lucyleeow**
#### What does this implement/fix? Explain your changes.
Currently `_has_screen` is run every few seconds, during status checks, and is (most likely) causing AWS worker to hang. 

https://github.com/paris-saclay-cds/ramp-board/blob/7e975d01c080278b2101a5ca48461bc7e5dcd3ea/ramp-engine/ramp_engine/aws/api.py#L713-L721

AWS worker hangs (>50% of the time) with the last log being:
> 2020-10-14 21:37:49,621 RAMP-AWS DEBUG ssh -o 'StrictHostKeyChecking no' -i /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/aws_ramp_ireland.pem ubuntu@34.242.228.66 "screen -ls|awk '{print $1}' | cut -d. -f2 | grep submission_000000022| wc -l"

which is generally run ~20 or more times.

This uses the dispatcher config ~`check_finished_training_interval_secs`:~ `status_check_wait`
to add a wait time between status checks. The default wait time is 1s.

TODO:

* Add test
* Add documentation

### CODE REVIEW

**Comment by lucyleeow** [comment]
I am not sure this is the solution. Below are logs using the latest commit:

> 2020-10-16 20:20:37,498 RAMP-AWS DEBUG ssh -o 'StrictHostKeyChecking no' -i /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/aws_ramp_ireland.pem ubuntu@52.19.4.175 "screen -ls|awk '{print $1}' | cut -d. -f2 | grep submission_000000027| wc -l"
2020-10-16 20:20:42,387 RAMP-AWS DEBUG ssh -o 'StrictHostKeyChecking no' -i /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/aws_ramp_ireland.pem ubuntu@52.19.4.175 "screen -ls|awk '{print $1}' | cut -d. -f2 | grep submission_000000027| wc -l"
2020-10-16 20:20:47,193 RAMP-AWS DEBUG ssh -o 'StrictHostKeyChecking no' -i /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/aws_ramp_ireland.pem ubuntu@52.19.4.175 "screen -ls|awk '{print $1}' | cut -d. -f2 | grep submission_000000027| wc -l"
2020-10-16 20:20:52,511 RAMP-AWS DEBUG ssh -o 'StrictHostKeyChecking no' -i /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/aws_ramp_ireland.pem ubuntu@52.19.4.175 "screen -ls|awk '{print $1}' | cut -d. -f2 | grep submission_000000027| wc -l"
2020-10-16 20:20:56,030 RAMP-AWS INFO check_finished_training_interval_secs: dt:5.310669 min_wait:60
2020-10-16 20:21:56,087 RAMP-AWS INFO Collecting submission 'submission_000000027'
2020-10-16 20:21:57,720 RAMP-AWS DEBUG rsync -e "ssh -o 'StrictHostKeyChecking no' -i /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/aws_ramp_ireland.pem" -avzP ubuntu@52.19.4.175:/home/ubuntu/ramp-kits/human_locomotion/submissions/submission_000000027/log /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/events/human_locomotion_aws/logs/submission_000000027/log
2020-10-16 20:22:04,045 RAMP-AWS DEBUG ssh -o 'StrictHostKeyChecking no' -i /home/lucy/Documents/RAMP/ramp-board/ramp-deployment/aws_ramp_ireland.pem ubuntu@52.19.4.175 "ls -l /home/ubuntu/ramp-kits/human_locomotion/submissions/submission_000000027/training_output|grep fold_|wc -l"

There are 24 `screen -ls`... calls done before our logger info of 'check_finished_training_interval_secs: dt:5.310669 min_wait:60'. I am not sure where this call is done. Here:

https://github.com/paris-saclay-cds/ramp-board/blob/7e975d01c080278b2101a5ca48461bc7e5dcd3ea/ramp-engine/ramp_engine/aws/api.py#L105-L107

? But the logger text 'Training of submission' is not present in the logs at all.

cc @agramfort @glemaitre @maikia

**Comment by lucyleeow** [comment]
ping @glemaitre @agramfort ready for review

File: ramp-engine/ramp_engine/dispatcher.py [hunk 1]
```diff
@@ -174,14 +178,23 @@ def collect_result(self, session):
             elif self.hunger_policy == 'exit':
                 self._poison_pill = True
             return
+
         for worker, (submission_id, submission_name) in zip(workers,
                                                             submissions):
+            dt = worker.time_since_last_status_check()
+            wait_longer = (False if dt is None else
+                           dt < self.status_check_wait)
+            if wait_longer:
```
Comment (glemaitre) [inline ramp-engine/ramp_engine/dispatcher.py:187]:
```suggestion
            dt = worker.time_since_last_status_check()
            if dt is not None and dt < self.time_between_collection:
```

File: ramp-engine/ramp_engine/dispatcher.py [hunk 2]
```diff
@@ -174,14 +178,23 @@ def collect_result(self, session):
...
+                self._processing_worker_queue.put_nowait(
+                    (worker, (submission_id, submission_name)))
+                time.sleep(0)
+                continue
             if worker.status == 'running':
```
Comment (glemaitre) [inline ramp-engine/ramp_engine/dispatcher.py:192]: I think that we can safely put an `elif` here

```suggestion
            elif worker.status == 'running':
```

**Review by glemaitre — COMMENTED** [review]
It looks good. I made some comment to change the name of the variable mainly.
Could you also add the new parameter to the template config file:

https://github.com/paris-saclay-cds/ramp-board/blob/master/ramp-utils/ramp_utils/template/ramp_config_template.yml

In this way, it will be obvious when deploying even if someone does not read the documentation first.

**Comment by lucyleeow** [comment]
Thanks for the review @glemaitre ! Changes made.

cc @agramfort

**Review by glemaitre — APPROVED** [review]
LGTM on my side. I will draft a test when I find a bit of free time.

**Comment by lucyleeow** [comment]
I've tested on my local RAMP server (using an AWS worker) 3 times, making 3 submissions at the same time, used both `ramp dispatcher` and the daemon and it has run fine.

I will note that it would be safe to put `n_workers` as one less than our limit. When I tried to run 4 submissions at the same time I got this error:

> botocore.exceptions.ClientError: An error occurred (VcpuLimitExceeded) when calling the RunInstances operation: You have requested more vCPU capacity than your current vCPU limit of 24 allows for the instance bucket that the specified instance type belongs to. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.

Full traceback:
> Traceback (most recent call last):
  File "/home/lucy/miniconda3/envs/ramp/bin/ramp-launch", line 11, in <module>
    load_entry_point('ramp-engine', 'console_scripts', 'ramp-launch')()
  File "/home/lucy/Documents/RAMP/ramp-board/ramp-engine/ramp_engine/cli.py", line 122, in start
    main()
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/lucy/Documents/RAMP/ramp-board/ramp-engine/ramp_engine/cli.py", line 90, in dispatcher
    disp.launch()
  File "/home/lucy/Documents/RAMP/ramp-board/ramp-engine/ramp_engine/dispatcher.py", line 278, in launch
    self.launch_workers(session)
  File "/home/lucy/Documents/RAMP/ramp-board/ramp-engine/ramp_engine/dispatcher.py", line 156, in launch_workers
    worker.setup()
  File "/home/lucy/Documents/RAMP/ramp-board/ramp-engine/ramp_engine/aws/worker.py", line 63, in setup
    self.instance, = aws.launch_ec2_instances(self.config)
  File "/home/lucy/Documents/RAMP/ramp-board/ramp-engine/ramp_engine/aws/api.py", line 150, in launch_ec2_instances
    SecurityGroups=[security_group],
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(**params)
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/lucy/miniconda3/envs/ramp/lib/python3.7/site-packages/botocore/client.py", line 661, in _make_api_call
    raise error_class(parsed_response, operation_name)

**Comment by lucyleeow** [comment]
@agramfort I will make a bug release but won't update on the server until next week as I know you have your course.

**Comment by lucyleeow** [comment]
@agramfort I'll leave the button to you.

**Comment by agramfort** [comment]
thx @lucyleeow !

**Pull request merged**